### PR TITLE
Optimize LightGCN Model

### PR DESCRIPTION
**Pull request by darrylong**
### Description

Optimizing LightGCN model to make it consistent with NGCF model.

### Checklist:

- [ ] I have added tests.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated `README.md` (if you are adding a new model).
- [ ] I have updated `examples/README.md` (if you are adding a new example).
- [ ] I have updated `datasets/README.md` (if you are adding a new dataset).

### CODE REVIEW

File: cornac/models/lightgcn2/lightgcn.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            }
+        )
+
+    def forward(self, g, users=None, pos_items=None, neg_items=None):
+        h_dict = {ntype: self.feature_dict[ntype] for ntype in g.ntypes}
+        # obtain features of each layer and concatenate them all
+        user_embeds = h_dict[USER_KEY]
+        item_embeds = h_dict[ITEM_KEY]
+
+        for i, layer in enumerate(self.layers):
```
Comment (tqtg) [inline cornac/models/lightgcn2/lightgcn.py:103]: maybe we can change `i` to `k` to match the notion in the paper?

File: cornac/models/lightgcn2/recom_lightgcn.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+            # we will use numpy for faster prediction in the score function, no need torch
+            self.U = u_embs.cpu().detach().numpy()
+            self.V = i_embs.cpu().detach().numpy()
+
+            if self.early_stopping is not None and self.early_stop(
+                **self.early_stopping
+            ):
+                break
+
+    def monitor_value(self):
```
Comment (tqtg) [inline cornac/models/lightgcn2/recom_lightgcn.py:193]: I think we'll be using recall@20 to do early stopping similar to NGCF instead of the loss value

File: cornac/models/lightgcn2/recom_lightgcn.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    num_epochs: int, default: 1000
+        Maximum number of iterations or the number of epochs.
+
+    learning_rate: float, default: 0.001
+        The learning rate that determines the step size at each iteration
+
+    train_batch_size: int, default: 1024
+        Mini-batch size used for train set
+
+    test_batch_size: int, default: 100
```
Comment (tqtg) [inline cornac/models/lightgcn2/recom_lightgcn.py:43]: this will ignored once early stopping changed to monitoring recall@20

File: examples/lightgcn_example.py [hunk 4]
```diff
@@ -33,6 +33,18 @@
     rating_threshold=0.5,
 )
 
+lightgcn1 = cornac.models.LightGCN1(
+    name="LightGCN-old",
+    seed=123,
+    num_epochs=2000,
+    num_layers=3,
+    early_stopping={"min_delta": 1e-4, "patience": 3},
```
Comment (tqtg) [inline examples/lightgcn_example.py:41]: I guess we can make the hyper-parameters similar to the example of NGCF?

**Review by tqtg — CHANGES_REQUESTED** [review]

File: examples/lightgcn_example.py (hunk 4 above)
Comment (darrylong) [reply examples/lightgcn_example.py:41]: Actually it already closely follows NGCF, just that 'num_layers' instead of 'layer_sizes' because we dont need to specify feature size for each layer.

Comment (tqtg) [reply examples/lightgcn_example.py:41]: Based on [this](https://github.com/PreferredAI/cornac/blob/master/examples/ngcf_example.py), there are a few things we can align:
```python
num_epochs=1000
early_stopping={"min_delta": 1e-4, "patience": 50}
train_batch_size=1024
lambda_reg=1e-5
```

Comment (darrylong) [reply examples/lightgcn_example.py:41]: Oh yes, you are right. But for lambda_reg I think in lightGCN paper they mentioned 1e-4 is optimal. Should we keep that, or follow NGCF?

Comment (tqtg) [reply examples/lightgcn_example.py:41]: We can keep 1e-4 as in the paper

**Comment by tqtg** [comment]
@darrylong shall we try to clean up this PR and merge it?

File: cornac/models/lightgcn/recom_lightgcn.py [hunk 5]
```diff
@@ -93,13 +84,11 @@ def __init__(
         seed=2020,
     ):
         super().__init__(name=name, trainable=trainable, verbose=verbose)
-
+        self.emb_size = emb_size
         self.num_epochs = num_epochs
         self.learning_rate = learning_rate
-        self.hidden_dim = hidden_dim
-        self.num_layers = num_layers
         self.train_batch_size = train_batch_size
```
Comment (tqtg) [inline cornac/models/lightgcn/recom_lightgcn.py:93]: Shall we shorten it to `batch_size`?

**Comment by tqtg** [comment]
LGTM

**Review by tqtg — APPROVED** [review]
